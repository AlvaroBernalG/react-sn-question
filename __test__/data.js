export const mockUpTest = {
  id: 33,
  title: "An animal cell contains:",
  questions: [
    {
      id: 0,
      correct: 0,
      selected: 1,
      options: [
        {
          name: "Cell Wall",
          value: "Cell Wall",
          id: 1
        },
        {
          name: "Ribosomes",
          value: "Ribosomes",
          id: 2
        }
      ]
    },
    {
      id: 1,
      correct: 0,
      selected: 0,
      options: [
        {
          name: "Cytoplasm",
          value: "Cytoplasm",
          id: 3
        },
        {
          name: "Chloroplast",
          value: "Chloroplast",
          id: 4
        }
      ]
    },
    {
      id: 2,
      correct: 0,
      selected: 0,
      options: [
        {
          name: "Impermeable membrane",
          value: "Impermeable membrane",
          id: 5
        },
        {
          name: "Partially permeable membrane",
          value: "Partially permeable membrane",
          id: 6
        }
      ]
    }
  ]
};
