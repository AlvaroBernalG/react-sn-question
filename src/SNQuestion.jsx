import React from "react";
import PropTypes from "prop-types";
import Switch, { State } from "react-switchable";
import "react-switchable/dist/main.css";
import "./SNQuestion.scss";

export default class SNQuestion extends React.Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
    resolutionMessage: PropTypes.string.isRequired,
    questions: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        correct: PropTypes.number.isRequired,
        options: PropTypes.arrayOf(
          PropTypes.shape({
            name: PropTypes.string.isRequired,
            value: PropTypes.string.isRequired
          })
        ),
        selected: PropTypes.number
      })
    ).isRequired,
    onChange: PropTypes.func,
    classStates: PropTypes.arrayOf(PropTypes.string), // eslint-disable-line react/no-unused-prop-types
    disable: PropTypes.bool
  };

  static defaultProps = {
    onChange: undefined,
    classStates: undefined,
    disable: false
  };

  static defaultClassStates = ["failure", "hot", "medium", "almost", "success"];

  static computeScore = questions =>
    questions.reduce(
      (acc, question) => acc + (question.selected === question.correct ? 1 : 0),
      0
    );

  static updateQuestions = (questions, { questionIndex, answerIndex }) =>
    questions.map(
      (question, index) =>
        questionIndex === index
          ? { ...question, selected: answerIndex }
          : question
    );

  static isQuestionCorrect = questions =>
    questions.every(q => q.selected === q.correct);

  static computeDefaultClasses = (classes, nthQuestions) => [
    classes[0],
    ...classes.slice(1, nthQuestions),
    classes[classes.length - 1]
  ];

  static getDerivedStateFromProps(props, prevState) {
    // if props.classStates is being defined from the parent
    // initialize state with the data.
    if (props.classStates) {
      return {
        classStates: props.classStates
      };
    }
    // if classes has not been defined from parent and the state has not
    // been initialized:
    if (prevState.classStates.length === 0) {
      return {
        classStates: SNQuestion.computeDefaultClasses(
          SNQuestion.defaultClassStates,
          props.questions.length
        )
      };
    }
    return null;
  }

  state = {
    classStates: []
  };

  getMainWrapperClasses(questions) {
    const computedScore = SNQuestion.computeScore(questions);
    const finalClass = this.state.classStates[computedScore];
    const classes = ["snquestion", `snquestion--${finalClass}`]
      .join(" ")
      .trim();
    return classes;
  }

  changeHandler(selectedValue, answerIndex, questionIndex) {
    if (this.props.onChange) {
      this.props.onChange(selectedValue, answerIndex, questionIndex);
    }
  }

  render() {
    const { questions, title, disable, resolutionMessage } = this.props;
    return (
      <div className={this.getMainWrapperClasses(questions)}>
        <h1 className="snquestion__title">{title}</h1>
        {questions.map((question, answerIndex) => (
          <Switch
            key={question.id}
            disable={disable}
            onValueChange={(value, questionIndex) =>
              this.changeHandler(value, questionIndex, answerIndex)
            }
          >
            {question.options.map((option, optionIndex) => (
              <State
                active={question.selected === optionIndex}
                key={option.id}
                value={option.name}
              >
                {option.value}
              </State>
            ))}
          </Switch>
        ))}
        <p className="snquestion__result">{resolutionMessage}</p>
      </div>
    );
  }
}
