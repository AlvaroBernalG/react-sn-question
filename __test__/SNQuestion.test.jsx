import React from "react";
import { mount } from "enzyme";
import SNQuestion from "../src/";
import { mockUpTest } from "./data";

const defaultProps = { ...mockUpTest,  resolutionMessage: "works" } 

const wrapper = props => <SNQuestion {...props} />;

describe("<SNQuestion />", () => {
  it("should correctly mount the component.", () => {
    const enzWrapper = mount(wrapper(defaultProps));
    expect(enzWrapper.find(".snquestion").length).toBe(1);
  });

  it("computeScore() should correctly compute the score.", () => {
    expect(SNQuestion.computeScore(mockUpTest.questions)).toBe(2);
  });

  it("updateQuestions() should correctly update questions.", () => {
    const updatedQuestions = SNQuestion.updateQuestions(mockUpTest.questions, {
      questionIndex: 0,
      answerIndex: 0
    });
    expect(updatedQuestions[0].selected).toBe(0);
  });

  it("isQuestionCorrect() should return true if all answers are correct.", () => {
    const updatedQuestions = SNQuestion.updateQuestions(mockUpTest.questions, {
      questionIndex: 0,
      answerIndex: 0
    });
    expect(SNQuestion.isQuestionCorrect(updatedQuestions)).toBe(true);
  });

  it("should correctly render the class states.", () => {
    const enzwrapper = mount(wrapper(defaultProps));
    enzwrapper.setProps({ classStates: ["one", "two", "three"] });
    expect(enzwrapper.state("classStates").join("")).toBe(
      ["one", "two", "three"].join("")
    );
    
    expect(enzwrapper.find(".snquestion--three").length).toBe(1);
  });

  it("should correctly render the title / footer.", () => {
    const enzwrapper = mount(wrapper(defaultProps));
    enzwrapper.setProps({ resolutionMessage: "works" });
    expect(enzwrapper.find(".snquestion__result").text()).toBe("works");
    enzwrapper.setProps({ title: "works" });
    expect(enzwrapper.find(".snquestion__title").text()).toBe("works");
  });
});
