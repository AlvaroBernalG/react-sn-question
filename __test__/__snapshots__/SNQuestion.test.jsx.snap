// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SNQuestion /> should match the snapshot. 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SNQuestion
    classStates={undefined}
    disable={false}
    id={33}
    onChange={undefined}
    questions={
      Array [
        Object {
          "correct": 0,
          "id": 0,
          "options": Array [
            Object {
              "id": 1,
              "name": "Cell Wall",
              "value": "Cell Wall",
            },
            Object {
              "id": 2,
              "name": "Ribosomes",
              "value": "Ribosomes",
            },
          ],
          "selected": 1,
        },
        Object {
          "correct": 0,
          "id": 1,
          "options": Array [
            Object {
              "id": 3,
              "name": "Cytoplasm",
              "value": "Cytoplasm",
            },
            Object {
              "id": 4,
              "name": "Chloroplast",
              "value": "Chloroplast",
            },
          ],
          "selected": 0,
        },
        Object {
          "correct": 0,
          "id": 2,
          "options": Array [
            Object {
              "id": 5,
              "name": "Impermeable membrane",
              "value": "Impermeable membrane",
            },
            Object {
              "id": 6,
              "name": "Partially permeable membrane",
              "value": "Partially permeable membrane",
            },
          ],
          "selected": 0,
        },
      ]
    }
    resolutionMessage="works"
    title="An animal cell contains:"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="snquestion__title"
        >
          An animal cell contains:
        </h1>,
        Array [
          <SNToggles
            changeHandler={[Function]}
            correct={0}
            disable={false}
            id={0}
            options={
              Array [
                Object {
                  "id": 1,
                  "name": "Cell Wall",
                  "value": "Cell Wall",
                },
                Object {
                  "id": 2,
                  "name": "Ribosomes",
                  "value": "Ribosomes",
                },
              ]
            }
            selected={1}
            switchIndex={0}
          />,
          <SNToggles
            changeHandler={[Function]}
            correct={0}
            disable={false}
            id={1}
            options={
              Array [
                Object {
                  "id": 3,
                  "name": "Cytoplasm",
                  "value": "Cytoplasm",
                },
                Object {
                  "id": 4,
                  "name": "Chloroplast",
                  "value": "Chloroplast",
                },
              ]
            }
            selected={0}
            switchIndex={1}
          />,
          <SNToggles
            changeHandler={[Function]}
            correct={0}
            disable={false}
            id={2}
            options={
              Array [
                Object {
                  "id": 5,
                  "name": "Impermeable membrane",
                  "value": "Impermeable membrane",
                },
                Object {
                  "id": 6,
                  "name": "Partially permeable membrane",
                  "value": "Partially permeable membrane",
                },
              ]
            }
            selected={0}
            switchIndex={2}
          />,
        ],
        <p
          className="snquestion__result"
        >
          works
        </p>,
      ],
      "className": "snquestion snquestion--medium",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "An animal cell contains:",
          "className": "snquestion__title",
        },
        "ref": null,
        "rendered": "An animal cell contains:",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "changeHandler": [Function],
          "correct": 0,
          "disable": false,
          "id": 0,
          "options": Array [
            Object {
              "id": 1,
              "name": "Cell Wall",
              "value": "Cell Wall",
            },
            Object {
              "id": 2,
              "name": "Ribosomes",
              "value": "Ribosomes",
            },
          ],
          "selected": 1,
          "switchIndex": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "changeHandler": [Function],
          "correct": 0,
          "disable": false,
          "id": 1,
          "options": Array [
            Object {
              "id": 3,
              "name": "Cytoplasm",
              "value": "Cytoplasm",
            },
            Object {
              "id": 4,
              "name": "Chloroplast",
              "value": "Chloroplast",
            },
          ],
          "selected": 0,
          "switchIndex": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "changeHandler": [Function],
          "correct": 0,
          "disable": false,
          "id": 2,
          "options": Array [
            Object {
              "id": 5,
              "name": "Impermeable membrane",
              "value": "Impermeable membrane",
            },
            Object {
              "id": 6,
              "name": "Partially permeable membrane",
              "value": "Partially permeable membrane",
            },
          ],
          "selected": 0,
          "switchIndex": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "works",
          "className": "snquestion__result",
        },
        "ref": null,
        "rendered": "works",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="snquestion__title"
          >
            An animal cell contains:
          </h1>,
          Array [
            <SNToggles
              changeHandler={[Function]}
              correct={0}
              disable={false}
              id={0}
              options={
                Array [
                  Object {
                    "id": 1,
                    "name": "Cell Wall",
                    "value": "Cell Wall",
                  },
                  Object {
                    "id": 2,
                    "name": "Ribosomes",
                    "value": "Ribosomes",
                  },
                ]
              }
              selected={1}
              switchIndex={0}
            />,
            <SNToggles
              changeHandler={[Function]}
              correct={0}
              disable={false}
              id={1}
              options={
                Array [
                  Object {
                    "id": 3,
                    "name": "Cytoplasm",
                    "value": "Cytoplasm",
                  },
                  Object {
                    "id": 4,
                    "name": "Chloroplast",
                    "value": "Chloroplast",
                  },
                ]
              }
              selected={0}
              switchIndex={1}
            />,
            <SNToggles
              changeHandler={[Function]}
              correct={0}
              disable={false}
              id={2}
              options={
                Array [
                  Object {
                    "id": 5,
                    "name": "Impermeable membrane",
                    "value": "Impermeable membrane",
                  },
                  Object {
                    "id": 6,
                    "name": "Partially permeable membrane",
                    "value": "Partially permeable membrane",
                  },
                ]
              }
              selected={0}
              switchIndex={2}
            />,
          ],
          <p
            className="snquestion__result"
          >
            works
          </p>,
        ],
        "className": "snquestion snquestion--medium",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "An animal cell contains:",
            "className": "snquestion__title",
          },
          "ref": null,
          "rendered": "An animal cell contains:",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "changeHandler": [Function],
            "correct": 0,
            "disable": false,
            "id": 0,
            "options": Array [
              Object {
                "id": 1,
                "name": "Cell Wall",
                "value": "Cell Wall",
              },
              Object {
                "id": 2,
                "name": "Ribosomes",
                "value": "Ribosomes",
              },
            ],
            "selected": 1,
            "switchIndex": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "changeHandler": [Function],
            "correct": 0,
            "disable": false,
            "id": 1,
            "options": Array [
              Object {
                "id": 3,
                "name": "Cytoplasm",
                "value": "Cytoplasm",
              },
              Object {
                "id": 4,
                "name": "Chloroplast",
                "value": "Chloroplast",
              },
            ],
            "selected": 0,
            "switchIndex": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "changeHandler": [Function],
            "correct": 0,
            "disable": false,
            "id": 2,
            "options": Array [
              Object {
                "id": 5,
                "name": "Impermeable membrane",
                "value": "Impermeable membrane",
              },
              Object {
                "id": 6,
                "name": "Partially permeable membrane",
                "value": "Partially permeable membrane",
              },
            ],
            "selected": 0,
            "switchIndex": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "works",
            "className": "snquestion__result",
          },
          "ref": null,
          "rendered": "works",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
